pkg/provenance.go

```diff
@@ -393,8 +395,137 @@ func VerifyProvenance(env *dsselib.Envelope, expectedHash string) error {
 	}
 
 	if !strings.EqualFold(hash, expectedHash) {
-		return errorMismatchHash
+		return fmt.Errorf("hash '%s': %w", hash, errorMismatchHash)
 	}
 
 	return nil
 }
+
+func VerifyBranch(env *dsselib.Envelope, expectedBranch string) error {
+	branch, err := getBranch(env)
+	if err != nil {
+		return err
+	}
+
+	if !strings.EqualFold(branch, "refs/heads/"+expectedBranch) {
+		return fmt.Errorf("branch '%s': %w", branch, errorMismatchBranch)
+	}
+
+	return nil
+}
+
+func getAsInt(parameters map[string]interface{}, field string) (int, error) {
+	value, ok := parameters[field]
+	if !ok {
+		return -1, fmt.Errorf("%w: %s", errorInvalidDssePayload,
+			fmt.Sprintf("parameters type for %s", field))
+	}
+
+	i, ok := value.(float64)
+	if !ok {
+		return -1, fmt.Errorf("%w: %s", errorInvalidDssePayload, "parameters type float64")
+	}
+	return int(i), nil
+}
+
+func getAsString(parameters map[string]interface{}, field string) (string, error) {
+	value, ok := parameters[field]
+	if !ok {
+		return "", fmt.Errorf("%w: %s", errorInvalidDssePayload,
+			fmt.Sprintf("parameters type for %s", field))
+	}
+
+	i, ok := value.(string)
+	if !ok {
+		return "", fmt.Errorf("%w: %s", errorInvalidDssePayload, "parameters type string")
+	}
+	return i, nil
+}
+
+func getBaseRef(parameters map[string]interface{}) (string, error) {
+	baseRef, err := getAsString(parameters, "base_ref")
+	if err != nil {
+		return "", err
+	}
+
+	// This `base_ref` seems to always be "".
+	if baseRef != "" {
+		return baseRef, nil
+	}
+
+	// Look at the event payload instead.
+	// We don't do thatt it all the time because the payload
+	// is event-specific; and only the `push` event seems to have a `base_ref``.
+	eventName, err := getAsString(parameters, "event_name")
+	if err != nil {
+		return "", err
+	}
+
+	if eventName != "push" {
+		return "", nil
+	}
+
+	eventPayload, ok := parameters["event_payload"]
+	if !ok {
+		return "", fmt.Errorf("%w: %s", errorInvalidDssePayload, "parameters type event payload")
+	}
+
+	payload, ok := eventPayload.(map[string]interface{})
+	if !ok {
+		return "", fmt.Errorf("%w: %s", errorInvalidDssePayload, "parameters type payload")
+	}
+
+	bRef, ok := payload["base_ref"]
+	if !ok {
+		return "", fmt.Errorf("%w: %s", errorInvalidDssePayload, "parameters type base_ref")
+	}
+
+	baseRef, ok = bRef.(string)
```

Comment:
can you `getAsString` the baseRef?

Reply:
that does not work. getAsString needs a map, the input is just a string. Right?

Reply:
getAsString(payload, 'base_ref')?

Reply:
done, thanks!